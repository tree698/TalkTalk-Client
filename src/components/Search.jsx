import React, { useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { BsSearch } from 'react-icons/bs';
import { RxCross2 } from 'react-icons/rx';
import { RxDividerVertical } from 'react-icons/rx';

export default function Search() {
  const [searchTerm, setSearchTerm] = useState('');
  const navigate = useNavigate();

  const handleChange = (e) => {
    const term = e.currentTarget.value;
    setSearchTerm((prev) => term);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (!searchTerm) {
      window.alert('Please, enter search term!');
      return;
    }
    navigate('/home/searcheddrawings', { state: { searchTerm } });
  };
  return (
    <form onSubmit={handleSubmit}>
      <div className="flex justify-between text-xs md:text-sm lg:text-base py-1 md:py-2 border border-superLightGray rounded-md shadow-inner">
        <input
          type="text"
          onChange={handleChange}
          value={searchTerm}
          placeholder="Search..."
          className="outline-none placeholder:text-black pl-2 md:pl-3 lg:pl-4"
        />
        <div className="flex items-center pr-1 md:pr-2 lg:pr-3">
          <span
            onClick={() => setSearchTerm((prev) => '')}
            className="text-xs md:text-sm lg:text-lg hover:scale-125 transition-all delay-150 duration-300 ease-in-out"
          >
            <RxCross2 />
          </span>
          <RxDividerVertical className="text-xs md:text-sm lg:text-lg text-lightGray" />
          <button className="hover:scale-125 transition-all delay-150 duration-300 ease-in-out">
            <BsSearch />
          </button>
        </div>
      </div>
    </form>
  );
}
