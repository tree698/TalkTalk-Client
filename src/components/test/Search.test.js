import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Route, useLocation } from 'react-router-dom';
import renderer from 'react-test-renderer';
import { withRouter } from '../../test/utils';
import Search from '../Search';

describe('Search', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      withRouter(<Route path="/" element={<Search />} />)
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  describe('Form Submit', () => {
    it('display the search term when typed a search term', () => {
      render(withRouter(<Route path="/" element={<Search />} />));
      const input = screen.getByPlaceholderText('Search...');

      userEvent.type(input, 'test');
      const inputElement = screen.getByRole('textbox');
      const inputValue = inputElement.value;
      expect(inputValue).toBe('test');
    });

    it('navigate to searcheddrawings with searchTerm when the submit button is clicked', () => {
      function LocationStateDisplay() {
        return <pre>{JSON.stringify(useLocation().state)}</pre>;
      }
      render(
        withRouter(
          <>
            <Route path="/" element={<Search />} />
            <Route
              path={`/home/searcheddrawings`}
              element={<LocationStateDisplay />}
            />
          </>
        )
      );
      const input = screen.getByPlaceholderText('Search...');
      const button = screen.getByRole('button');

      userEvent.type(input, 'test');
      userEvent.click(button);

      expect(
        screen.getByText(JSON.stringify({ searchTerm: 'test' }))
      ).toBeInTheDocument();
    });

    it('display warning message when clicking the submit button without searchTerm ', () => {
      const mockAlert = jest
        .spyOn(window, 'alert')
        .mockImplementation(() => {});

      function LocationStateDisplay() {
        return <pre>{JSON.stringify(useLocation().state)}</pre>;
      }

      render(
        withRouter(
          <>
            <Route path="/" element={<Search />} />
            <Route
              path={`/home/searcheddrawings`}
              element={<LocationStateDisplay />}
            />
          </>
        )
      );
      const button = screen.getByRole('button');

      userEvent.click(button);

      expect(mockAlert).toHaveBeenCalled();
      mockAlert.mockRestore();
    });

    it('delete searchTerm when clicking delete btn', () => {
      render(withRouter(<Route path="/" element={<Search />} />));
      const deleteBtn = screen.getByTitle('delete');
      const inputElement = screen.getByRole('textbox');
      const inputValue = inputElement.value;

      userEvent.click(deleteBtn);

      expect(inputValue).toBe('');
    });
  });
});
