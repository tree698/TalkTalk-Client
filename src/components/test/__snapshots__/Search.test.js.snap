// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search renders correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    className="flex justify-between text-xs md:text-sm lg:text-base py-1 md:py-2 border border-superLightGray rounded-md shadow-inner"
  >
    <input
      className="outline-none placeholder:text-black pl-2 md:pl-3 lg:pl-4"
      onChange={[Function]}
      placeholder="Search..."
      type="text"
      value=""
    />
    <div
      className="flex items-center pr-1 md:pr-2 lg:pr-3"
    >
      <span
        className="text-xs md:text-sm lg:text-lg hover:scale-125 transition-all delay-150 duration-300 ease-in-out"
        onClick={[Function]}
        title="delete"
      >
        <svg
          fill="none"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 15 15"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clipRule="evenodd"
            d="M11.7816 4.03157C12.0062 3.80702 12.0062 3.44295 11.7816 3.2184C11.5571 2.99385 11.193 2.99385 10.9685 3.2184L7.50005 6.68682L4.03164 3.2184C3.80708 2.99385 3.44301 2.99385 3.21846 3.2184C2.99391 3.44295 2.99391 3.80702 3.21846 4.03157L6.68688 7.49999L3.21846 10.9684C2.99391 11.193 2.99391 11.557 3.21846 11.7816C3.44301 12.0061 3.80708 12.0061 4.03164 11.7816L7.50005 8.31316L10.9685 11.7816C11.193 12.0061 11.5571 12.0061 11.7816 11.7816C12.0062 11.557 12.0062 11.193 11.7816 10.9684L8.31322 7.49999L11.7816 4.03157Z"
            fill="currentColor"
            fillRule="evenodd"
          />
        </svg>
      </span>
      <svg
        className="text-xs md:text-sm lg:text-lg text-lightGray"
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 15 15"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clipRule="evenodd"
          d="M7.5 2C7.77614 2 8 2.22386 8 2.5L8 12.5C8 12.7761 7.77614 13 7.5 13C7.22386 13 7 12.7761 7 12.5L7 2.5C7 2.22386 7.22386 2 7.5 2Z"
          fill="currentColor"
          fillRule="evenodd"
        />
      </svg>
      <button
        className="hover:scale-125 transition-all delay-150 duration-300 ease-in-out"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 16 16"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M11.742 10.344a6.5 6.5 0 1 0-1.397 1.398h-.001c.03.04.062.078.098.115l3.85 3.85a1 1 0 0 0 1.415-1.414l-3.85-3.85a1.007 1.007 0 0 0-.115-.1zM12 6.5a5.5 5.5 0 1 1-11 0 5.5 5.5 0 0 1 11 0z"
          />
        </svg>
      </button>
    </div>
  </div>
</form>
`;
